lib/manager/components/LabelEditor.js

```diff
@@ -0,0 +1,236 @@
+// @flow
+
+import React from 'react'
+import {
+  Button,
+  Col,
+  ControlLabel,
+  FormControl,
+  FormGroup,
+  HelpBlock,
+  ListGroup,
+  ListGroupItem,
+  Panel,
+  Row
+} from 'react-bootstrap'
+import update from 'react-addons-update'
+import { connect } from 'react-redux'
+
+import { createLabel, updateLabel } from '../../manager/actions/labels'
+import { validationState } from '../util'
+import type { ManagerUserState } from '../../types/reducers'
+import type { Label } from '../../types'
+
+type Props = {
+  createLabel: Function,
+  label: Label,
+  onDone?: Function,
+  projectId: string,
+  updateLabel: Function,
+  user: ManagerUserState
+}
+
+type State = {
+  labelIsNew?: boolean,
+  newLabel: Label,
+  validation: {
+    description: boolean,
+    name: boolean
+  }
+}
+
+class LabelEditor extends React.Component<Props, State> {
+    state = {
+      validation: {
+        name: true,
+        description: true
+      },
+      newLabel: {...this.props.label},
+      labelIsNew: false
+    }
+
+    componentDidMount = () => {
+      // If we didn't get a label passed in, declare the label to be new
+      if (Object.keys(this.props.label).length === 0) {
+        this.setState({
+          labelIsNew: true,
+          // Default to a nice orange
+          // we can safely set a partial object here
+          // $FlowFixMe
+          newLabel: {
+            color: '#FF9200'
+          }
+        })
+      }
+    }
+
+    userIsAdmin = (user: ManagerUserState) => {
+      const {permissions} = user
+      if (!permissions) return false
+      else return permissions.isApplicationAdmin() || permissions.canAdministerAnOrganization()
+    }
+
+  _onFormChange = ({target}: {target: HTMLInputElement}, alwaysValid: boolean) => {
+    const {name, value, type, checked} = target
+
+    const universalValue: string | boolean = type === 'checkbox' ? checked : value
+    // extra check for flow
+    const valid =
+      alwaysValid ||
+      (typeof universalValue === 'string' && universalValue.length > 0)
+
+    this.setState(
+      update(this.state, {
+        newLabel: { $merge: {[name]: universalValue} },
+        validation: { [name]: { $set: universalValue && valid } }
+      })
+    )
+  }
+
+  // compares the dirty label from the old one to determine
+  // which keys changed
+  _getChanges = () => {
+    const newLabel = this.state.newLabel
+    const oldLabel = this.props.label
+
+    const changes: any = {}
+    Object.keys(newLabel).map(k => {
```

Comment:
`.forEach`

Reply:
My bad for copying from elsewhere in the code. Refactored!